code/game/objects/structures/xeno.dm

```diff
@@ -728,109 +723,130 @@ TUNNEL
+		return
+
+	repair_xeno_structure(M) //Repair if possible
+
+	if(charges >= 5)
+		to_chat(M, "<span class='xenodanger'>[src] is already full!</span>")
+		return
+	if(ccharging)
+		to_chat(M, "<span class='xenodanger'>[src] is already being filled!</span>")
+		return
+
+	if(M.plasma_stored < XENO_ACID_WELL_FILL_COST) //You need to have enough plasma to attempt to fill the well
+		to_chat(M, "<span class='xenodanger'>We don't have enough plasma to fill [src]! We need [XENO_ACID_WELL_FILL_COST - M.plasma_stored] more plasma!</span>")
+		return
+
+	ccharging = TRUE
+	to_chat(M, "<span class='xenodanger'>We begin refilling [src]...</span>")
+	if(!do_after(M, XENO_ACID_WELL_FILL_TIME, FALSE, src, BUSY_ICON_BUILD))
 		ccharging = FALSE
-		update_icon()
-		to_chat(M,"<span class='xenonotice'>We fill up by one [src].</span>")
-	else
-		to_chat(M, "<span class='xenowarning'>We begin removing [src]...</span>")
-		if(do_after(M, 5 SECONDS, FALSE, src, BUSY_ICON_BUILD))
-			deconstruct(FALSE)
+		to_chat(M, "<span class='xenodanger'>We abort refilling [src]!</span>")
 		return
 
-/obj/effect/alien/resin/acidwell/Crossed(atom/A)
+	if(M.plasma_stored < XENO_ACID_WELL_FILL_COST)
+		ccharging = FALSE
+		to_chat(M, "<span class='xenodanger'>We don't have enough plasma to fill [src]! We need [XENO_ACID_WELL_FILL_COST - M.plasma_stored] more plasma!</span>")
+		return
+
+	M.plasma_stored -= XENO_ACID_WELL_FILL_COST
+	charges++
+	ccharging = FALSE
+	update_icon()
+	to_chat(M,"<span class='xenonotice'>We add acid to [src]. It is currently has [charges] / [XENO_ACID_WELL_MAX_CHARGES] charges.</span>")
+
+
+/obj/structure/acidwell/Crossed(atom/A)
 	. = ..()
 	if(iscarbon(A))
 		HasProximity(A)
 
-/obj/effect/alien/resin/acidwell/HasProximity(atom/movable/AM)
-	if(!iscarbon(AM))
+/obj/structure/acidwell/HasProximity(atom/movable/AM)
+	if(!isliving(AM))
 		return
-	var/mob/living/carbon/C = AM
-	if(C.stat == DEAD)
+	var/mob/living/stepper = AM
+	if(stepper.stat == DEAD)
 		return
 	if(!charges)
 		return
-	if(isxeno(C))
-		if(!(C.on_fire))
+	if(isxeno(stepper))
+		if(!(stepper.on_fire))
 			return
-		C.ExtinguishMob()
+		stepper.ExtinguishMob()
 		charges--
 		update_icon()
 		return
 	else
 		if(!charges)
 			return
-		C.adjustToxLoss(charges * 15)
+		stepper.apply_damage(charges * 30, BURN, BODY_ZONE_PRECISE_L_FOOT, stepper.run_armor_check(BODY_ZONE_PRECISE_L_FOOT, "acid"))
+		stepper.apply_damage(charges * 30, BURN, BODY_ZONE_PRECISE_R_FOOT, stepper.run_armor_check(BODY_ZONE_PRECISE_R_FOOT, "acid"))
+		stepper.visible_message("<span class='danger'>[stepper] is immersed in [src]'s acid!</span>", \
+		"<span class='danger'>We are immersed in [src]'s acid!</span>", null, 5)
+		playsound(stepper, "sound/bullets/acid_impact1.ogg", 10 * charges)
+		new /obj/effect/temp_visual/acid_bath(get_turf(stepper))
+		var/datum/effect_system/smoke_spread/xeno/acid/A = new(get_turf(stepper))
+		A.set_up(1,src)
+		A.start()
 		charges = 0
 		update_icon()
 		return
```

Comment:
early return and excessive returns

Reply:
well, that's how it was, but it's easy to fix